crypto/crypto/src/multi_ed25519.rs

```diff
@@ -0,0 +1,547 @@
+    fn try_from(bytes: &[u8]) -> std::result::Result<MultiEd25519PublicKey, CryptoMaterialError> {
+        let payload_length = bytes.len();
+        // Special case for single keys to speed up deserialization.
+        if payload_length == ED25519_PUBLIC_KEY_LENGTH {
+            return Ed25519PublicKey::try_from(bytes)
+                .map(|public_key| MultiEd25519PublicKey(vec![public_key], 1u8));
+        }
+        let threshold = get_threshold(
+            payload_length,
+            ED25519_PRIVATE_KEY_LENGTH,
+            bytes[payload_length - 1],
+        )?;
+
+        let mut public_keys: Vec<Ed25519PublicKey> = vec![];
+        let multi_key = bytes
+            .chunks(ED25519_PUBLIC_KEY_LENGTH)
+            .filter(|chunk| chunk.len() == ED25519_PUBLIC_KEY_LENGTH)
+            .try_for_each(|key_bytes| deserialize_and_push(key_bytes, &mut public_keys));
+
+        multi_key.map(|_| MultiEd25519PublicKey(public_keys, threshold))
+    }
+}
+
+/// We deduce VerifyingKey from pointing to the signature material
+/// we get the ability to do `pubkey.validate(msg, signature)`
+impl VerifyingKey for MultiEd25519PublicKey {
+    type SigningKeyMaterial = MultiEd25519PrivateKey;
+    type SignatureMaterial = MultiEd25519Signature;
+}
+
+impl fmt::Display for MultiEd25519PublicKey {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{}", hex::encode(&self.to_bytes()))
+    }
+}
+
+impl fmt::Debug for MultiEd25519PublicKey {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "MultiEd25519PublicKey({})", self)
+    }
+}
+
+impl Length for MultiEd25519PublicKey {
+    fn length(&self) -> usize {
+        self.to_bytes().len()
+    }
+}
+
+impl ValidKey for MultiEd25519PublicKey {
+    fn to_bytes(&self) -> Vec<u8> {
+        self.to_bytes()
+    }
+}
+
+impl MultiEd25519Signature {
+    /// This method will also sort signatures based on index.
+    pub fn new(
+        signatures: Vec<(Ed25519Signature, u8)>,
+    ) -> std::result::Result<Self, CryptoMaterialError> {
+        let num_of_sigs = signatures.len();
+        if num_of_sigs == 0 || num_of_sigs > MAX_NUM_OF_KEYS {
+            return Err(CryptoMaterialError::WrongLengthError);
+        }
+
+        let mut sorted_signatures = signatures.clone();
+        sorted_signatures.sort_by(|a, b| a.1.cmp(&b.1));
+
+        let mut sigvec: Vec<Ed25519Signature> = Vec::with_capacity(num_of_sigs);
+        let mut bitvec = BitVec::from_elem(MAX_NUM_OF_KEYS, false);
+
+        // Check if all indexes are unique and < MAX_NUM_OF_KEYS
+        for (i, item) in sorted_signatures.iter().enumerate() {
+            // If an index is out of range.
+            if item.1 < MAX_NUM_OF_KEYS as u8 {
+                // if an index has been set already (thus, there is a duplicate).
+                if bitvec[i] {
+                    return Err(CryptoMaterialError::FormatError);
```

Comment:
Good point @sblackshear, maybe we can use the `WrongBitVecError` in lines 318, 324 and 458 as well. Unless we keep the FormatError everywhere, without specifying the reason. Another option is to always return an error String info along with the type. Wdyt?

Reply:
Agreed that WrongBitVec fits nicely here + elsewhere! Error string wouldn't be bad either, but it looks like it might not be consistent with the way the rest of this API does errors.